File: python/E4.py
import threading
import queue
import random
def p(q,lock):
    lock.acquire()
    i = random.randint(1,100)
    q.put(random.randint(1,100))
    print(threading.currentThread().getName(),"p",i)
    lock.release()
def c(q,lock):
    r = 0
    while (not r):
        lock.acquire()
        if (not q.empty()):
            r = 1
            print(threading.currentThread().getName(),"c",q.get())
        lock.release()
t = []
q = queue.Queue()
lock = threading.Lock()
for i in range(1,12):
    if (i %2):
        t.append(threading.Thread(None, p,i,(q,lock)))
    else:
        t.append(threading.Thread(None,c,i,(q,lock)))

for i in t:
    i.start()

for i in t:
    i.join()

File: java/E4.java

public class E4 implements Runnable{

    public void run() {
        try {
            Thread.sleep(Long.MAX_VALUE);
        } catch (InterruptedException e) {
            System.out.println(("error :" + Thread.currentThread().getName()));
        }
        while(!Thread.interrupted());
        System.out.println(Thread.currentThread().getName());
    }

    public static void main(String[] args) throws InterruptedException{
        Thread t1 = new Thread(new E4(), "t1");
        t1.start();
        System.out.println( Thread.currentThread().getName() + "sleeps 3s");
        Thread.sleep(3000);
        System.out.println(Thread.currentThread().getName() + " interrupted");
        t1.interrupt();
        System.out.println( Thread.currentThread().getName() + "sleeps 3s");
        Thread.sleep(3000);
        System.out.println( Thread.currentThread().getName() + " interrupted");
        t1.interrupt();
    }

}